design/026-requirements_traits.md

```diff
@@ -0,0 +1,125 @@
+# Proposal: Define requirements new traits to qualify and enhance the Conan dependencies model
+
+
+| **Status**        |                                                   |
+|:------------------|:--------------------------------------------------|
+| **RFC #**         | [026](https://github.com/conan-io/tribe/pull/26)  |
+| **Submitted**     | 2021-07-06                                        |
+| **Tribe votes**   |                                                   |
+
+
+## Summary
+
+Introduce the concept of "requirements traits", to better model the relations between packages. Requirement traits are specifiers of the Conan ``requires`` that can be defined in recipes. These traits can allow defining the current Conan 1.X ``build_requires`` behavior with some ``requires`` traits.
+
+This proposal does not affect ``python_requires``, because of their nature and evaluation at recipe parsing time, they are very different to regular ``requires`` and ``build_requires``, and they would not benefit from this proposal.
+
+Requirement traits will define things like headers inclusions, library linkage, runtime dependency, if the headers visibility should propagate downstream, if the require should propagate downstream (and can conflict) or not, if it is a build tool or a test library, etc.
+
+Package recipes will keep defining their direct dependencies in the same way they were doing in Conan 1.X, but with the possibility of specifying more traits (might not be necessary in many cases, the defaults should be fine). But recipes will have access, once the graph is evaluated, to the ``requires`` of all their dependencies, both direct(declared) and indirect ones (transitive), describing the relation they have with those dependencies with these traits.
+
+
+## Motivation
+
+Conan's main declaration of dependencies is the ``requires`` relationship. Mostly 2 variants of it are also available in Conan 1.X: ``build_requires`` that aim to express a dependency to a build tool, like cmake, and ``private`` which is discouraged for most cases.
+
+But this model does not provide enough information to describe some specifics of the C/C++ compilation model, for example in the general case, when one library is linking another library, the consumers of the former shouldn’t have visibility over the headers of the later one. A ``requires`` trait called ``transitive_headers`` could control this visibility.
+It seems the potential combinations of dependencies usage is too high to consider approaches like the current one, defining high level names as ``build_requires``, ``test_requires``, ``requires_shared``, etc. does not scale.
+
+
+## Detailed Design
+
+The proposed traits are arguments to the ``requires``, like:
+
+```python
+def requirements(self):
+    self.requires(ref, headers=True, libs=True, run=None, visible=True, …)
```

Comment:
Is the extension to ref as tuple also planned?

Would it be possible to write something like

```python
class MyRecipe(ConanFile):
    ...
    requires = (
        'pkg1/1.0.0',
        ('gtest/1.10.0', 'test'),
        ('pkg2/2.0.0', 'visible=False,package_id=True')
    )
```

or something similar?

Currently it's possible to provide only a single string `override` or `private`.

Reply:
From the patterns I see both from ConanCenter and customers, there is clear trend to use more and more ``def requirements()`` method. So I'd suggest:

- Leave the plain ``requires = "zlib/1.2.11"`` for the standard library requirement case, with all defaults
- For anything that deviates from the defaults, need to use the method ``requirements()`` form, as it provides better attribute checking, is more idiomatic in general.

Reply:
Since this proposal will create a discrepancy in behavior/functionality... would it be possible to remove (or deprecate) the attribute mechanism? I'd love to avoid the confusion of two similarly name workflows and which when to use when and having to write hooks or best practices.

Reply:
I would assume the point of keeping the attribute form defined is to still have a simple, purely declarative format to that can be defined equivalently in conanfile.py and conanfile.txt. 

Though this proposal certainly does undermine that by adding new consumer-oriented kwargs which could only accessed by a conanfile.py.